### ISSUE TITLE
Add an API that combines autocomplete with arbitrary cypher queries

### ISSUE THREAD

**Issue by cmungall** [SciGraph/SciGraph#177]
Background: The current SciGraph API breaks down functionalities into separate areas: lexical, annotation, graph. This is a useful high level breakdown but frequently we need to combine these. This is one thing the old oboedit tool got right that has never been duplicated in Protege, the ability to combine advanced lexical queries with graph queries (cc @dosumis).

In this particular case, we would like to be able to filter auto-complete results based on a cypher query. A simple but common example is an autocomplete field that is restricted to be only transitive reflexive subclasses of a given ontology class, where the members are not in a given subset. See https://github.com/geneontology/noctua/issues/295 for example. A more sophisticated example would use taxon constraints: e.g. restrict this autocomplete to classes applicable for mammalia.

Because this is potentially expensive we may want to restrict the set of cypher queries that can be used to something pre-defined by the SciGraph admin, specified using a yaml file analogous to the current `dynamic` option. The yaml would specify the name of the autocomplete (e.g. 'biological process mammalia') and map this to a cypher query. There could also be a hint as to whether the cypher should be applied after the search to filter out entries, or prior to the search, to create a whitelist.

I like the idea of a power api user being able to specify arbitrary cypher but don't know the implications in terms of killer queries.

**Comment by DoctorBud** [comment]
Another use case for this capability is in the ObOne browser experiment that Mark P and I have been working with. Specifically, we are currently using the `/graph/neighbors` endpoint to get all neighbors of distance 1 from the currently active node. However, this endpoint has no ability to have paged results (e.g., `count`, `startIndex` and `limit` parameters). It would be nice to be able to give the user the ability to filter these neighbors based on the same criteria used for autocomplete, so that as the user types into a *Filter By* field, the set of neighbors seen is pruned to match the filter.

So basically, an optional parameter to /autocomplete that indicated that candidates should be neighbors of distance=1 would be useful. Alternatively, adding a filterBy expression to /neighbors would work.

**Comment by cmungall** [comment]
On 2 Jun 2016, at 6:22, Dan Keith wrote:

> Another use case for this capability is in the ObOne browser 
> experiment that Mark P and I have been working with. Specifically, we 
> are currently using the `/graph/neighbors` endpoint to get all 
> neighbors of distance 1 from the currently active node. However, this 
> endpoint has no ability to have paged results (e.g., `count`, 
> `startIndex` and `limit` parameters).

sounds like it would be useful to have this


> It would be nice to be able to give the user the ability to filter 
> these neighbors based on the same criteria used for autocomplete, so 
> that as the user types into a *Filter By* field, the set of neighbors 
> seen is pruned to match the filter.

I'm not really sure about the utility of this, as to do anything useful 
that isn't hardcoded to any one use case you'd need the full power of a 
query language like cypher

But for the text filtering case, you could piggy back this off the 
autocomplete call, but that feels an odd mixing of concerns
